unittests/svnn_finality_violation_tests.cpp

```diff
@@ -0,0 +1,758 @@
+            ("from", "eosio"_n)
+            ("to", "user1"_n)
+            ("quantity", "100.0000 EOS")
+            ("memo", "");
+
+        mutable_variant_object user1_transfer = mvo()
+            ("from", "user1"_n)
+            ("to", "user2"_n)
+            ("quantity", "1.0000 EOS")
+            ("memo", "");
+
+        //setup a light client data cache
+
+        //data retained for every policy change
+        struct policy_sunset_data { 
+            
+            finalizer_policy policy;
+
+            ibc_block_data_t last_qc_block;
+            qc_data_t last_qc;
+
+        };
+
+        struct data_cache {
+
+            //The data_cache stores data relevant to finality violation proofs.
+            //It only operates in optimistic mode, which is sufficient for finality violation proofs testing purposes
+
+            //store the reversible blocks
+            std::vector<finality_block_data> reversible_blocks;
+
+            //store the last block over which we have a QC, as well as said QC
+            finality_block_data high_qc_block;
+            qc_t high_qc;
+
+            //store the last final block, as wall as the first QC over it. The last_final_qc and high_qc together constitute the 2-chains required for finality progress
+            finality_block_data last_final_block;
+            qc_t last_final_qc;
+
+            //store all policies sunset data
+            std::vector<policy_sunset_data> policy_sunsets;
+
+            //current active finalizer policy
+            finalizer_policy active_finalizer_policy;
+
+            //observe a stream of blocks as they are received, and store minimal data required to construct finality violation proofs in the future
+            ibc_block_data_t scan_block(const ibc_block_data_t& block_result){
+
+                finality_block_data block_data{
+                    .block_num = block_result.block->block_num(),
+                    .timestamp = block_result.block->timestamp,
+                    .parent_timestamp = block_result.parent_timestamp,
+                    .finality_digest = block_result.finality_digest,
+                    .last_pending_finalizer_policy_start_timestamp = block_result.last_pending_finalizer_policy_start_timestamp,
+                    .level_3_commitments = block_result.level_3_commitments,
+                    .base_digest = block_result.base_digest,
+                    .finality_root = block_result.finality_root,
+                    .qc_data = block_result.qc_data
+                };
+
+                if (block_result.qc_data.qc.has_value()) {
+
+                    auto high_qc_block_itr = std::find_if(reversible_blocks.begin(), reversible_blocks.end(), [&](auto& r) {
+                        return r.block_num == block_result.qc_data.qc.value().to_qc_claim().block_num;
+                    });
+
+                    if ( high_qc_block_itr != reversible_blocks.end()
+                         && high_qc_block_itr->qc_data.qc.has_value()) {
```

Comment:
Since the test is design to run "optimistically", where every block contains a new strong QC, this check and the following one are only here to produce the correct behavior for the first few blocks which don't have a QC, in which case we do not need to continue searching the reversible blocks.